src/Data/Group/Free.hs

```diff
@@ -139,18 +147,71 @@ present = flip ($)
 -- The intuition here is group elements correspond with their positive
 -- or negative multiplicities, and as such are simplified by construction.
 --
-newtype FreeAbelianGroup a = FreeAbelianGroup { runFreeAbelian :: Map a Int }
-    deriving (Show, Eq, Ord)
+-- === __Examples__:
+-- 
+-- >>> a = singleton 'a'
+-- >>> b = singleton 'b'
+-- >>> a
+-- FreeAbelianGroup $ fromList [('a',1)]
+-- >>> a <> b
+-- FreeAbelianGroup $ fromList [('a',1),('b',1)]
+-- >>> a <> b == b <> a
+-- True
+-- >>> invert a
+-- FreeAbelianGroup $ fromList [('a',-1)]
+-- >>> a <> b <> invert a
+-- FreeAbelianGroup $ fromList [('b',1)]
+-- >>> gtimes 5 (a <> b)
+-- FreeAbelianGroup $ fromList [('a',5),('b',5)]
+newtype FreeAbelianGroup a = MkFreeAbelianGroup (Map a Integer)
+    deriving (Eq, Ord)
+
+instance Show a => Show (FreeAbelianGroup a) where
+    showsPrec p (MkFreeAbelianGroup g) =
+        showParen (p > 0) $ ("FreeAbelianGroup $ " ++) . shows g
+
+-- | /O(n)/ Constructs a 'FreeAbelianGroup' from a finite 'Map' from
+--   the set of generators (@a@) to its multiplicities.
+makeFreeAbelianGroup :: Ord a => Map a Integer -> FreeAbelianGroup a
+makeFreeAbelianGroup = MkFreeAbelianGroup . Map.filter (/= 0)
+
+-- | /O(1)/ Gets a representation of 'FreeAbelianGroup' as
+--   'Map'. The returned map contains no records with
+--   multiplicity @0@ i.e. @'Map.lookup' a@ on the returned map
+--   never returns @Just 0@.
+runFreeAbelianGroup :: FreeAbelianGroup a -> Map a Integer
+runFreeAbelianGroup (MkFreeAbelianGroup g) = g
+
+pattern FreeAbelianGroup :: Ord a => Map a Integer -> FreeAbelianGroup a
+pattern FreeAbelianGroup g <- MkFreeAbelianGroup g where
+    FreeAbelianGroup g = makeFreeAbelianGroup g
+
+{-# COMPLETE FreeAbelianGroup #-}
 
 instance (Ord a) => Semigroup (FreeAbelianGroup a) where
-    (FreeAbelianGroup g) <> (FreeAbelianGroup g') =
-      FreeAbelianGroup $ Map.unionWith (+) g g'
+    (MkFreeAbelianGroup g) <> (MkFreeAbelianGroup g') =
+      MkFreeAbelianGroup $ mergeG g g'
+      where
+        mergeG = Map.merge
+          Map.preserveMissing
+          Map.preserveMissing
+          (Map.zipWithMaybeMatched $ \_ m n -> nonZero $ m + n)
+        nonZero n = if n == 0 then Nothing else Just n
+    
+    stimes = gtimes
```

Comment:
may as well just use `pow` here and avoid the indirection

Reply:
```suggestion
    stimes = flip pow
```